src/aiida_quantumespresso/workflows/pw/base.py

```diff
@@ -405,21 +406,34 @@ def handle_known_unrecoverable_failure(self, calculation):
         exit_codes=[
             PwCalculation.exit_codes.ERROR_COMPUTING_CHOLESKY,
             PwCalculation.exit_codes.ERROR_DIAGONALIZATION_TOO_MANY_BANDS_NOT_CONVERGED,
+            PwCalculation.exit_codes.ERROR_S_MATRIX_NOT_POSITIVE_DEFINITE,
+            PwCalculation.exit_codes.ERROR_ZHEGVD_FAILED,
+            PwCalculation.exit_codes.ERROR_QR_FAILED,
+            PwCalculation.exit_codes.ERROR_EIGENVECTOR_CONVERGENCE,
+            PwCalculation.exit_codes.ERROR_BROYDEN_FACTORIZATION,
         ]
     )
     def handle_diagonalization_errors(self, calculation):
         """Handle known issues related to the diagonalization.
 
-        Switch to ``diagonalization = 'cg'`` if not already running with this setting, and restart from the charge
-        density. In case the run already used conjugate gradient diagonalization, abort.
+        We use the following strategy. When a diagonalization algorithm fails, we try using an other one
+        still not used. Conjugate gradient (CG) is kept as last option, as it is the slowest among the
+        available ones, but on the contrary it is the most stable as well, thus kept as `last resort`.
+
+        If also CG is failing, abort.
         """
-        if self.ctx.inputs.parameters['ELECTRONS'].get('diagonalization', None) == 'cg':
+        current = self.ctx.inputs.parameters['ELECTRONS'].get('diagonalization', 'david')
+
+        if current == 'cg':
```

Comment:
Is a failure with `cg` truly hopeless? No chance that e.g. running with `paro` might work? We could just treat `cg` on an equal footing as the other diagonalization settings, keeping it last in the order of preference.

Reply:
Yes, one can definitely try and see if any of the available ones is working

Reply:
Simple trial and error, love it. :D Frankly I don't remember a case where `david` failed and `paro` didn't fix the issue, so it's probably a bit overkill, but 🤷‍♂️

Reply:
Right, forgot about reason for the ordering - no one in particular, but for some cases (magnetic ones), I found `ppcg` to be more stable. But again, anything better than `cg`